treewide: Fix passthru.updateScripts of NGI-maintained derivations

OPNA2608 opened the pull request:
When running the `passthru.updateScript`s of packages that I maintain (`nix-shell maintainers/scripts/update.nix --argstr maintainer OPNA2608`), I want the updating process to finish. If it stops halfway through due to an error, then I need to re-run the entire thing. So I want `passthru.updateScript`s to behave well enough that they won't run into a critical error and abort the entire procedure.

So here are some changes to make the `updateScript`s no longer misbehave.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ethancedwards8 commented on pkgs/by-name/ga/galene/package.nix:
```diff
@@ -37,9 +41,35 @@ buildGoModule (finalAttrs: {
+
+        fname="$(nix-instantiate --eval -E "with import ./. {}; (builtins.unsafeGetAttrPos \"version\" $UPDATE_NIX_ATTR_PATH).file" | cut -d'"' -f2)"
+
+        ${sd.meta.mainProgram} --string-mode "$oldhash" "$newhash" "$fname"
```
I think we use `lib.getExe` instead nowadays?

OPNA2608 replied:
Just dropped the "smart" meta lookup and hardcoded `sd`, more consistent with the rest of the script(s).

ethancedwards8 commented on pkgs/by-name/hy/hypercore/package.nix:
```diff
@@ -24,7 +30,60 @@ buildNpmPackage (finalAttrs: {
+        fi
+
+        # For NPM to behave nice
+        HOME="$(mktemp -d)"
```
same as above?

ethancedwards8 commented on pkgs/by-name/hy/hyperswarm/package.nix:
```diff
@@ -24,7 +30,60 @@ buildNpmPackage (finalAttrs: {
+        fi
+
+        # For NPM to behave nice
+        HOME="$(mktemp -d)"
```
above

ethancedwards8 commented on pkgs/by-name/au/autobase/package.nix:
```diff
@@ -27,7 +33,60 @@ buildNpmPackage (finalAttrs: {
+        fi
+
+        # For NPM to behave nice
+        HOME="$(mktemp -d)"
```
Could this be changed to `writableTmpDirAsHomeHook` in `runtimeInputs` instead? Not sure.

OPNA2608 replied:
`updateScript`s don't run as a derivation with a strong sandbox or phases or anything AFAIU, they're kinda just executables that get invoked with some special environment variables set, and within your writable copy of Nixpkgs. I doubt that adding hook derivations works… Could maybe try to `source` the underlying script file, and then run `writableTmpDirAsHome`, but *is that really much better than a quick `mktemp -d`,,,*

ethancedwards8 reviewed (changes requested):
it seems like a lot of duplicate code/update scripts. Could they be consolidated/shared across each other? maybe a `npmUpdateScript`?

ethancedwards8 commented:
@mic92 curious to hear your thoughts on what is happening here as well...

Prince213 commented:
It would be better if we can look into why nix-update-script failed before replacing it with a big trunk of shell script.

Mic92 commented:
Regarding the failing nix-update:

Maybe we run into: https://github.com/Mic92/nix-update/pull/433

OPNA2608 commented:
> It would be better if we can look into why nix-update-script failed before replacing it with a big trunk of shell script.

Agree, the proper solution would be to figure out what's going wrong and to fix the underlying issue.

But as I said, I *do not care* how this is solved, I just *need* this stuff to work. And the short-term fix for me is to drop `nix-update(-script`) usage…

> Regarding the failing nix-update:
> 
> Maybe this is caused by: [Mic92/nix-update#433](https://github.com/Mic92/nix-update/pull/433)

Probably:

```
error: hash mismatch in fixed-output derivation '/nix/store/p0kp6w60lc31q67y4i3hn5mdkh0qkpzf-c680bec003e5f40175ef030a9b90c92cc2acc78a.patch.drv':
           likely URL: https://github.com/Mic92/nix-update/commit/c680bec003e5f40175ef030a9b90c92cc2acc78a.patch
            specified: sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
                  got: sha256-KUG4sPtk/igyava8LYMUjNx0HSuWH9ikGiAObULnAQ4=
        expected path: /nix/store/d0615dmx3mw6gvfd0pww6g57z1ib095f-c680bec003e5f40175ef030a9b90c92cc2acc78a.patch
             got path: /nix/store/wwg4kwld6k7721p5pnywvvr36bssgmji-c680bec003e5f40175ef030a9b90c92cc2acc78a.patch
```

> Nix now actually has an API to get hash mismatches, maybe we can use this instead (the JSON log output can now be written to Unix socket output).

"Now" meaning very recently, or "after Nix 2.3"? I use Lix system-wide.

Mic92 commented:
> > It would be better if we can look into why nix-update-script failed before replacing it with a big trunk of shell script.
> 
> Agree, the proper solution would be to figure out what's going wrong and to fix the underlying issue.
> 
> But as I said, I _do not care_ how this is solved, I just _need_ this stuff to work. And the short-term fix for me is to drop `nix-update(-script`) usage…
> 
> > Regarding the failing nix-update:
> > Maybe this is caused by: [Mic92/nix-update#433](https://github.com/Mic92/nix-update/pull/433)
> 
> Probably:
> 
> ```
> error: hash mismatch in fixed-output derivation '/nix/store/p0kp6w60lc31q67y4i3hn5mdkh0qkpzf-c680bec003e5f40175ef030a9b90c92cc2acc78a.patch.drv':
>            likely URL: https://github.com/Mic92/nix-update/commit/c680bec003e5f40175ef030a9b90c92cc2acc78a.patch
>             specified: sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
>                   got: sha256-KUG4sPtk/igyava8LYMUjNx0HSuWH9ikGiAObULnAQ4=
>         expected path: /nix/store/d0615dmx3mw6gvfd0pww6g57z1ib095f-c680bec003e5f40175ef030a9b90c92cc2acc78a.patch
>              got path: /nix/store/wwg4kwld6k7721p5pnywvvr36bssgmji-c680bec003e5f40175ef030a9b90c92cc2acc78a.patch
> ```
> 
> > Nix now actually has an API to get hash mismatches, maybe we can use this instead (the JSON log output can now be written to Unix socket output).
> 
> "Now" meaning very recently, or "after Nix 2.3"? I use Lix system-wide.

nix-update sets nix in the PATH, so this must be caused by an overlay.
In nix 2.30 there is an `json-log-path` option that allows for more precise checksum updates:
https://github.com/NixOS/nix/blob/ab7feb38987fc4ff627c34ce09b3cadd7a44148d/doc/manual/source/release-notes/rl-2.30.md

OPNA2608 commented:
> nix-update sets nix in the PATH, so this must be caused by an overlay.

I only use `nix.package = pkgs.lix`, no nixpkgs-wide overlay for that. If it's the daemon that's emitting the message and not the CLI, then that would do it though.

SuperSandro2000 commented on pkgs/by-name/au/autobase/package.nix:
```diff
@@ -27,7 +33,60 @@ buildNpmPackage (finalAttrs: {
+        fname="$(nix-instantiate --eval -E "with import ./. {}; (builtins.unsafeGetAttrPos \"version\" $UPDATE_NIX_ATTR_PATH).file" | cut -d'"' -f2)"
+
+        # Replace hash & lockfile
+        sd --string-mode "$oldhash" "$newhash" "$fname"
```
Just use gnused, no need to force build a simple rust rewrite of it.

SuperSandro2000 commented on pkgs/by-name/au/autobase/package.nix:
```diff
@@ -27,7 +33,60 @@ buildNpmPackage (finalAttrs: {
+        prefetch-npm-deps
+        sd
+      ];
+      text = ''
```
Those long scripts should probably be in an extra update script

OPNA2608 replied:
You mean a file? Sure, can do.

OPNA2608 commented on pkgs/development/python-modules/sipsimple/default.nix:
```diff
@@ -64,13 +65,13 @@ let
 buildPythonPackage rec {
   pname = "python3-sipsimple";
-  version = "5.3.3.2";
+  version = "5.3.3.2-mac";
```
This switches away from `nix-update-script`, so disregarded.

OPNA2608 committed:
autobase: Update lockfile as part of the updateScript

OPNA2608 committed:
python3Packages.msrplib: Fix version

So passthru.updateScript is happy.

OPNA2608 committed:
python3Packages.sat-tmp: Disable checkPhase

Nothing to check, just like the comment in the derivation said.

OPNA2608 committed:
python3Packages.sat-tmp: Drop auto-added passthru.updateScript

nix-update doesn't handle mercurial sources.

OPNA2608 committed:
python3Packages.urwid-satext: Drop auto-added passthru.updateScript

nix-update doesn't handle mercurial sources.

OPNA2608 committed:
python3Packages.sipsimple: Fix updateScript's detection of new versions

OPNA2608 committed:
python3Packages.highctidh: Switch to gitUpdater

Because nix-update tries to do a downgrade.

OPNA2608 committed:
hypercore: Update lockfile as part of the updateScript

OPNA2608 committed:
hyperswarm: Update lockfile as part of the updateScript

OPNA2608 committed:
dokieli: Replace nix-update-script with custom solution

nix-update choked on trying to update offlineCache.hash, because it doesn't update missing-hashes.json.

OPNA2608 committed:
misskey: Switch from nix-update-script to gitUpdater

nix-update can't handle the pnpm lockfile needing patching.
